### PR TITLE
make unknown features on `cargo add` more discoverable

### PR DESCRIPTION
**Pull request by Emilgardis**
When adding an unknown feature via `cargo add -F krate/feat`, it can be easy to miss the fact that the change failed.

This fixes that by showing the following output on fail

image

### CODE REVIEW

**Comment by epage** [comment]
It seems like there is more to discuss on the usability of the error message and how we should handle it; can we move the discussion to an issue as [recommended in our contributing information](https://github.com/rust-lang/cargo/blob/master/CONTRIBUTING.md)?

File: src/cargo/ops/cargo_add/mod.rs [hunk 1]
```diff
@@ -124,11 +126,71 @@ pub fn add(workspace: &Workspace<'_>, options: &AddOptions<'_>) -> CargoResult<(
...
-            anyhow::bail!("unrecognized features: {unknown_features:?}");
+            let (activated, deactivated) = dep.features();
+            let deactivated = deactivated
+                .iter()
+                .filter(|f| !unknown_features.contains(f))
+                .collect::<Vec<_>>();
+            let activated = activated
+                .iter()
+                .filter(|f| !unknown_features.contains(f))
+                .collect::<Vec<_>>();
```
Comment (epage) [inline src/cargo/ops/cargo_add/mod.rs:146]: Why is this needed?

File: src/cargo/ops/cargo_add/mod.rs [hunk 2]
```diff
@@ -124,11 +126,71 @@ pub fn add(workspace: &Workspace<'_>, options: &AddOptions<'_>) -> CargoResult<(
...
+                            })
+                            .into_iter()
+                            .format("\n    ")
+                    )?
+                }
+            }
+            anyhow::bail!(message.trim().to_owned());
         }
 
+        print_dep_table_msg(&mut options.config.shell(), &dep)?;
```
Comment (epage) [inline src/cargo/ops/cargo_add/mod.rs:192]: nit: ideally the splitting of the prints would be at least a separate commit from the new error message as these are related but separate changes.  If nothing else, this makes it easier on the reviewer.

File: src/cargo/ops/cargo_add/mod.rs [hunk 3]
```diff
@@ -124,11 +126,71 @@ pub fn add(workspace: &Workspace<'_>, options: &AddOptions<'_>) -> CargoResult<(
...
+                .filter(|f| !unknown_features.contains(f))
+                .collect::<Vec<_>>();
+            let mut message = format!(
+                "unrecognized feature{} for crate {}: {}\n",
+                if unknown_features.len() == 1 { "" } else { "s" },
+                dep.name,
+                unknown_features.iter().format(", "),
+            );
+            if activated.is_empty() && deactivated.is_empty() {
+                write!(message, "no features available for crate {}", dep.name)?;
```
Comment (epage) [inline src/cargo/ops/cargo_add/mod.rs:154]: We should probably have a test that covers this case as well

File: src/cargo/ops/cargo_add/mod.rs [hunk 4]
```diff
@@ -124,11 +126,71 @@ pub fn add(workspace: &Workspace<'_>, options: &AddOptions<'_>) -> CargoResult<(
...
             unknown_features.extend(inherited_features.difference(&available_features).copied());
         }
+
         unknown_features.sort();
+
+        let mut activated: IndexSet<_> =
+            dep.features.iter().flatten().map(|s| s.as_str()).collect();
+        if dep.default_features().unwrap_or(true) {
+            activated.insert("default");
+        }
```
Comment (epage) [inline src/cargo/ops/cargo_add/mod.rs:136]: Is this dead code?

File: src/cargo/ops/cargo_add/mod.rs (hunk 1 above)
Comment (Emilgardis) [reply src/cargo/ops/cargo_add/mod.rs:146]: if this is not included, the unknown features are included in the output.

```
$ cargo add tokio --features sync,net,rt,rr,dd
[...snip]
error: unrecognized features for crate tokio: dd, rr
disabled features:
    bytes, fs, full, io-std, io-util, macros, memchr, mio, num_cpus, parking_lot
    process, rt-multi-thread, signal, signal-hook-registry, stats, test-util, time
    tokio-macros, tracing, winapi
enabled features:
    dd, libc, net, once_cell, rr, rt, socket2, sync
```

I guess the filter on deactivated is not needed, is it possible to remove features with cargo add or will that come with the eventual `cargo remove/rm` merge?

File: src/cargo/ops/cargo_add/mod.rs (hunk 2 above)
Comment (Emilgardis) [reply src/cargo/ops/cargo_add/mod.rs:192]: good point, if you want I can still split it out

File: src/cargo/ops/cargo_add/mod.rs (hunk 4 above)
Comment (Emilgardis) [reply src/cargo/ops/cargo_add/mod.rs:136]: yes 👍🏼 if only clippy could be able to catch that case

File: src/cargo/ops/cargo_add/mod.rs (hunk 3 above)
Comment (Emilgardis) [reply src/cargo/ops/cargo_add/mod.rs:154]: I added this with `cargo_add::features_unknown::features_unknown_empty` not sure if it should really be added as it's own folder for `cargo_test`

File: src/cargo/ops/cargo_add/mod.rs (hunk 1 above)
Comment (Emilgardis) [reply src/cargo/ops/cargo_add/mod.rs]: I switched this to use `retain` instead

Comment (epage) [reply src/cargo/ops/cargo_add/mod.rs]: Oh because we've already merged them into the `DependencyUI`.  Probably deserves a comment.

File: src/cargo/ops/cargo_add/mod.rs (hunk 2 above)
Comment (epage) [reply src/cargo/ops/cargo_add/mod.rs:181]: I've already gone through the review of it here so not a blocker (would be if this were bigger) but up to you and for consideration for future changes.

File: src/cargo/ops/cargo_add/mod.rs (hunk 3 above)
Comment (epage) [reply src/cargo/ops/cargo_add/mod.rs:143]: Please use one test case per `mod.rs`.

**Comment by ehuss** [comment]
r? epage

**Comment by Emilgardis** [comment]
I think theres one more thing to do here, should probably clearly signal that the command failed, and that nothing was changed for the crate in the manifest

**Comment by epage** [comment]
@bors r+

> I think theres one more thing to do here, should probably clearly signal that the command failed, and that nothing was changed for the crate in the manifest

I think the PR is sufficient for now and we can always iterate further on it.

File: src/cargo/ops/cargo_add/mod.rs [hunk 5]
```diff
@@ -634,6 +688,42 @@ impl DependencyUI {
             })
             .collect();
     }
+
+    fn features(&self) -> (IndexSet<&str>, IndexSet<&str>) {
+        let mut activated: IndexSet<_> =
+            self.features.iter().flatten().map(|s| s.as_str()).collect();
+        if self.default_features().unwrap_or(true) {
+            activated.insert("default");
+        }
```
Comment (marcospb19) [inline src/cargo/ops/cargo_add/mod.rs:697]: Hi @Emilgardis! 
I'm trying to make a PR that touches this file and I wonder if this snippet is necessary, you add `"default"` to the activated features list and ignore it down below, can I remove it?

Comment (marcospb19) [reply src/cargo/ops/cargo_add/mod.rs:697]: oh, sorry, it is dead code indeed https://github.com/rust-lang/cargo/pull/11098#discussion_r974655861
